app/services/registration_checker.rb

```diff
@@ -89,6 +91,20 @@ def validate_create_events!
       raise RegistrationError.new(:forbidden, ErrorCodes::INVALID_EVENT_SELECTION) if event_limit.present? && event_ids.count > event_limit
     end
 
+    def validate_qualifications!
+      return unless @competition_info.enforces_qualifications?
+      # TODO: Read the request payload in as an object, and handle cases where expected values aren't found
+      event_ids = @request.dig('competing', 'event_ids')
+
+      unqualified_event = event_ids.find do |event|
+        qualification = @competition_info.get_qualification_for(event)
+        next if qualification.nil?
+        !competitor_qualifies_for_event?(event, qualification)
+      end
+
+      raise RegistrationError.new(:unprocessable_entity, ErrorCodes::QUALIFICATION_NOT_MET) if unqualified_event
```

Comment:
Why are you focussing on finding _one_ unqualified event? Wouldn't it make sense to report all unqualified events at once?

Otherwise competitors might go out to compete, think "Yay now I can finally register", come back to their PC, be happy that there's another Worlds spot available for grabs, and be disappointed that there is _yet another_ event they didn't qualify for but the system didn't tell them about previously.

Reply:
The premise here is that the frontend should be communicating with the user on what events they do/don't qualify for - ie, if a user doesn't meet qualification then the event should not be selectable.

Thus, the role of the backend is not to give the user detailed feedback on which qualifications aren't met, it's just to reject the registration outright as soon as it proves to be ineligible. 

I'm open to returning an array of events that the user doesn't meet qualification if you think that would be better, though.

Reply:
Disregard, we now return an array of unqualified events in a "data" parameter in the error payload